heartbeat/skima/core.go

```diff
@@ -0,0 +1,168 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package skima
+
+import (
+	"strings"
+
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"fmt"
+
+	"sort"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+// Is creates a named IsDef with the given checker.
+func Is(name string, checker Checker) IsDef {
+	return IsDef{name: name, checker: checker}
+}
+
+// Optional wraps an IsDef to mark the field's presence as optional.
+func Optional(id IsDef) IsDef {
+	id.name = "optional " + id.name
+	id.optional = true
+	return id
+}
+
+// Map is the type used to define schema definitions for Schema.
+type Map map[string]interface{}
+
+// SchemaValidator is the result of Schema and is run against the map you'd like to test.
+type SchemaValidator func(common.MapStr) SchemaValidationResults
```

Comment:
try to reduce stutter. The package kinda name resembles schema. For users of a package, a symbol defined in a package always has the name combined by package name + symbol name. E.g. the actual type name here is `skima.SchemaValidator`. Without stuttering we get `skima.Validator`.

Same for the validation results. The `SchemaValidationResults` will  be `skima.SchemaValidationResults`. Unless you have a more generic result type consider renaming this to `skima.Results`

Reply:
The confusing thing here is that there is a notion of Value validation vs. Schema validation in this package.

For values we have `IsDef`, `Checker`, and `ValueResult`
For map/schema validation we have `SchemaValidator` and `SchemaValidationResults`.

I'm thinking these names are a bit inconsistent, and I agree too long. WDYT about renaming them as indicated below?

`IsDef`->`IsDef`
`Checker`->`ValValidator`
`ValueResult`->`ValResult`
`SchemaValidator`->`Validator`
`SchemaValidationResults`->`Results`

Reply:
SGTM. `ValueResult` is fine as well.